### PR TITLE
Fix #334: Improve the batch norm algorithm

### LINKED ISSUE #334: Improve batchNorm
For discussing how to improve the [BatchNorm section](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-batchnorm). I am working on a draft PR.

There is a good note/example, and description of the parameters.

Formal steps are missing. They should spec how the input and output is handled. For instance, whether normalization can be done in-situ. May the impl modify the input and return the same object (I assume not, i.e. different JS objects, but may share underlying buffers).

In addition, I thought to add the batch norm algorithm from the linked paper as an informal, descriptive sub-steps section.

Also, somewhat more detailed introduction, including typical use cases (currently shortly mentioned).

Any inputs, examples, discussion welcome.

### PR DESCRIPTION
**Pull request by zolkis**
A draft PR for now. Needs review, though. 
Depends on prior merge of #337 and possibly #329.
Please make general comments and discussion in #334, and syntax related comments in this PR.

***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/webnn/pull/339.html" title="Last updated on May 16, 2023, 3:30 PM UTC (0eea13e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/339/9774d49...zolkis:0eea13e.html" title="Last updated on May 16, 2023, 3:30 PM UTC (0eea13e)">Diff</a>

### CODE REVIEW

**Comment by zolkis** [comment]
Fixed the issue of initialization vs execution for batch norm.
Included the layout tests from note to algorithmic steps.
Removed the statements about linking issues with #337. After that is merged, this PR will be re-checked and marked for review.

**Comment by zolkis** [comment]
Small changes done, removing WiP. Can be reviewed/merged. @huningxin PTAL.

File: index.bs [hunk 1]
```diff
@@ -1168,21 +1201,42 @@ partial interface MLGraphBuilder {
...
+        1. If |mean|.{{MLOperand/[[descriptor]]}}.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}} is not equal with |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}} from which the dimension represented by |options|.axis is removed, then  throw a "{{TypeError}}" {{DOMException}} and abort these steps.
+    1. Let |variance| be the third argument, representing the moving variance values of |input|.
+    1. Let |options| be the fourth argument. To validate |options|, run these substeps:
+        1. If |options|.axis does not [=map/exist=], let |options|."axis" be 1.
+        1. If |options|.axis is not a number between 0 and the rank of |input|, then throw a "{{TypeError}}" {{DOMException}} and abort these steps.
+        1. If |input| is a 4-D tensor of the *"nchw"* layout, set |options|.axis to 1.
+        1. If |input| is a 4-D tensor of the *"nhwc"* layout, set |options|.axis to 3.
+            <div class="note">
+                The axis value designates the feature or channel count dimension of the input tensor.
+            </div>
```
Comment (wchao1115) [inline index.bs:1233]: This is redundant to what already stated above this section.

File: index.bs [hunk 2]
```diff
@@ -1168,21 +1201,42 @@ partial interface MLGraphBuilder {
...
+    1. Let |result| be an {{MLOperand}} representing the results. It may use the same underlying data as |input|.
+    1. Issue a request to the underlying platform to initialize the batch normalization, given |result| to store the results and |options|. Wait for completion.
+        <div class="informalsteps">
+            1. If |options|.activation [=map/exists=], the implementation MAY use it to fuse it with the batch normalization.
+        </div>
```
Comment (wchao1115) [inline index.bs:1238]: If we feel compelled to add this note, there may be a better way to express it. The current wording doesn't read well because it doesn't explain what "fuse" in this context means, and what is "it" that the statement refers to. For example, a more complete wording may be "At the graph execution time, the implementation may choose to combine the execution of the standard batch normalization operation with the optional activation function provided by *options.activation* to improve the overall execution performance of the combined operation." etc.

**Review by wchao1115 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 2 above)
Comment (zolkis) [reply index.bs:1238]: I agree, but I think the use of "fuse" should be handled in a separate issue/PR, consistently over the whole spec.

**Comment by zolkis** [comment]
I have reset to the original text, and kept the new/moved text.
It can be re-reviewed. Thanks for the patience.

**Comment by zolkis** [comment]
@wchao1115 PTAL - reduced the branch to one clean commit.

**Comment by zolkis** [comment]
Requested changes have been done in https://github.com/webmachinelearning/webnn/pull/339/commits/0eea13ef3145cbd90f1dcd1ebd8bb8e74aff5e97 
Continued in / moved to #397 (changed the target branch from _main_ to 
https://github.com/webmachinelearning/webnn/tree/zk-conventions-integration

**Comment by zolkis** [comment]
Closing this PR, as it's moved to #397.
